README.md
# http-cache-control

HTTP `Cache-Control` header middleware for standard `Request` and `Response`.

## What

Middleware for HTTP `Content-Control` headers.

Assists in defining type-safe cache directives and safely adds `Cache-Control`
headers.

Compliant:

- [RFC 9111, 5.2. Cache-Control](https://www.rfc-editor.org/rfc/rfc9111.html#name-cache-control)
- [RFC 8246, HTTP Immutable Responses](https://www.rfc-editor.org/rfc/rfc8246.html)
- [RFC 5861, HTTP Cache-Control Extensions for Stale Content](https://www.rfc-editor.org/rfc/rfc5861)

## Middleware

For a definition of Universal HTTP middleware, see the
[http-middleware](https://github.com/httpland/http-middleware) project.

## Usage

Middleware factory is exported by default.

```ts
import cacheControl from "https://deno.land/x/http_cache_control@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

const middleware = cacheControl({
  "max-age": 604800,
  immutable: true,
  public: true,
});
declare const request: Request;
const handler = () => new Response("hello");

const response = await middleware(request, handler);

assertEquals(
  response.headers.get("cache-control"),
  "max-age=604800, immutable, public",
);
```

## Deep dive

If the `Response` already has a `Cache-Control` header, do nothing.

```ts
import cacheControl from "https://deno.land/x/http_cache_control@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

const middleware = cacheControl({
  "max-age": 604800,
  immutable: true,
  public: true,
});
declare const request: Request;
const handler = () =>
  new Response("hello", { headers: { "cache-control": "no-cache" } });

const response = await middleware(request, handler);

assertEquals(response.headers.get("cache-control"), "no-cache");
```

### Throwing error

If a number other than non-negative integer is specified, `AggregateError` will
be thrown.

```ts
import cacheControl from "https://deno.land/x/http_cache_control@$VERSION/mod.ts";
import { assertThrows } from "https://deno.land/std/testing/asserts.ts";

assertThrows(() => cacheControl({ "max-age": -100, "s-maxage": 1.1 }));
```

## License

Copyright © 2023-present [httpland](https://github.com/httpland).

Released under the [MIT](./LICENSE) license


CHANGELOG.md
# 1.0.0-beta.1 (2023-03-02)


### Features

* add cache-control header middleware factory ([289faf7](https://github.com/httpland/http-cache-control/commit/289faf7fed609608e632a8e3b7cea8da549e38f9))


mod_test.ts
import {
  assert,
  assertExists,
  assertThrows,
  describe,
  equalsResponse,
  it,
} from "./_dev_deps.ts";
import cacheControl from "./mod.ts";

describe("default export", () => {
  it("should exists", () => {
    assertExists(cacheControl);
  });

  it("should throw error", () => {
    assertThrows(() => cacheControl({ "max-age": -1, "s-maxage": -1.1 }));
  });

  it("should not throw error", () => {
    assert(cacheControl({ "max-age": 0, "s-maxage": 1 }));
  });

  it("should contain cache-control header", async () => {
    const middleware = cacheControl({
      "max-age": 604800,
      immutable: true,
      public: true,
    });
    const handler = () => new Response("hello");

    const response = await middleware(new Request("test:"), handler);

    assert(
      equalsResponse(
        response,
        new Response("hello", {
          headers: {
            "cache-control": "max-age=604800, immutable, public",
          },
        }),
      ),
    );
  });
});


util_test.ts
import { assertFalse, assertThrows, describe, it } from "./_dev_deps.ts";
import { assertValidDirectives } from "./util.ts";

describe("assertValidDirectives", () => {
  it("should throw error", () => {
    assertThrows(() =>
      assertValidDirectives({ "max-age": -1, "s-maxage": -1.1 })
    );
    assertThrows(() =>
      assertValidDirectives({ "max-age": 1.1, "s-maxage": 100 })
    );
  });

  it("should not throw error", () => {
    assertFalse(assertValidDirectives({ "max-age": 0, "s-maxage": 1 }));
    assertFalse(
      assertValidDirectives({
        "max-age": 60,
        immutable: true,
        public: true,
      }),
    );
  });
});


cache_control_test.ts
import {
  assert,
  assertEquals,
  describe,
  equalsResponse,
  it,
} from "./_dev_deps.ts";
import {
  CacheDirectives,
  ExtensionDirectives,
  stringifyCacheDirectives,
  stringifyDirective,
  withCacheControl,
} from "./cache_control.ts";

describe("stringifyDirective", () => {
  type Case = [string, number | true, string];

  it("should return empty if the key is empty", () => {
    const table: Case[] = [
      ["", 0, ""],
      ["", true, ""],
      ["", -1, ""],
    ];

    table.forEach(([key, value, expected]) => {
      assertEquals(stringifyDirective(key, value), expected);
    });
  });

  it("should return key-value format if the value is number", () => {
    const table: Case[] = [
      ["key", 0, "key=0"],
      ["key", -0, "key=0"],
      ["key", -1, "key=-1"],
    ];

    table.forEach(([key, value, expected]) => {
      assertEquals(stringifyDirective(key, value), expected);
    });
  });

  it("should return value only format if the value is true", () => {
    const table: Case[] = [
      ["key", true, "key"],
    ];

    table.forEach(([key, value, expected]) => {
      assertEquals(stringifyDirective(key, value), expected);
    });
  });
});

describe("stringifyCacheDirectives", () => {
  it("should return empty if the key is empty", () => {
    const table: [Record<string, number | true>, string][] = [
      [{}, ""],
      [{ key: true, key2: true }, "key, key2"],
      [{ key: 100, "key-2": true }, "key=100, key-2"],
    ];

    table.forEach(([record, expected]) => {
      assertEquals(stringifyCacheDirectives(record), expected);
    });
  });
});

describe("withCacheControl", () => {
  it("should return same response if the response has cache-control header yet", async () => {
    const table: [Request, Response, Response][] = [
      [
        new Request("test:"),
        new Response(null, { headers: { "cache-control": "" } }),
        new Response(null, { headers: { "cache-control": "" } }),
      ],
      [
        new Request("test:"),
        new Response(null, { headers: { "cache-control": "max-age=100" } }),
        new Response(null, { headers: { "cache-control": "max-age=100" } }),
      ],
    ];

    await Promise.all(table.map(async ([request, response, expected]) => {
      const result = await withCacheControl(
        request,
        response,
        { "max-age": 10000, immutable: true },
      );

      assert(equalsResponse(result, expected));
    }));
  });

  it("should add cache-control field", async () => {
    const table: [
      Request,
      Response,
      CacheDirectives & ExtensionDirectives,
      Response,
    ][] = [
      [
        new Request("test:"),
        new Response(null),
        {
          immutable: true,
          "max-age": 100,
          "must-revalidate": true,
          "must-understand": true,
          "no-cache": true,
          "no-store": true,
          "no-transform": true,
          private: true,
          "proxy-revalidate": true,
          public: true,
          "s-maxage": 1000,
          "stale-if-error": 1,
          "stale-while-revalidate": 10,
        },
        new Response(null, {
          headers: {
            "cache-control":
              "immutable, max-age=100, must-revalidate, must-understand, no-cache, no-store, no-transform, private, proxy-revalidate, public, s-maxage=1000, stale-if-error=1, stale-while-revalidate=10",
          },
        }),
      ],
    ];

    await Promise.all(
      table.map(async ([request, response, directives, expected]) => {
        const result = await withCacheControl(
          request,
          response,
          directives,
        );

        assert(await equalsResponse(result, expected, true));
      }),
    );
  });
});


mod.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

import { type Middleware } from "./deps.ts";
import {
  type CacheDirectives,
  type ExtensionDirectives,
  withCacheControl,
} from "./cache_control.ts";
import { assertValidDirectives } from "./util.ts";
export {
  type CacheDirectives,
  type ExtensionDirectives,
} from "./cache_control.ts";
export { type Middleware } from "./deps.ts";

/** HTTP `cache-control` header middleware factory.
 *
 * @example
 * ```ts
 * import cacheControl from "https://deno.land/x/http_cache_control@$VERSION/mod.ts";
 * import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
 *
 * const middleware = cacheControl({
 *   "max-age": 604800,
 *   immutable: true,
 *   public: true,
 * });
 * declare const request: Request;
 * const handler = () => new Response("hello");
 *
 * const response = await middleware(request, handler);
 *
 * assertEquals(
 *   response.headers.get("cache-control"),
 *   "max-age=604800, immutable, public",
 * );
 * ```
 *
 * @throws {AggregateError} If directives include invalid value.
 */
export default function cacheControl(
  directives: CacheDirectives & ExtensionDirectives,
): Middleware {
  assertValidDirectives(directives);

  return async (request, next) => {
    const response = await next(request);

    return withCacheControl(request, response, directives);
  };
}


util.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

import { filterValues, isNonNegativeInteger, isNumber } from "./deps.ts";
import {
  type CacheDirectives,
  type ExtensionDirectives,
} from "./cache_control.ts";
export {
  type CacheDirectives,
  type ExtensionDirectives,
} from "./cache_control.ts";
export { type Middleware } from "./deps.ts";

/** Assert the directives is valid.
 * @throws {AggregateError}
 */
export function assertValidDirectives(
  directives: CacheDirectives & ExtensionDirectives,
): asserts directives {
  const errorRecord = filterValues(
    { ...directives },
    (value) => isNumber(value) && !isNonNegativeInteger(value),
  );
  const keys = Object.keys(errorRecord);

  if (keys.length) {
    const errors = keys.map((key) =>
      RangeError(`directive must be non-negative integer. ${key}`)
    );
    throw AggregateError(errors, "directives include invalid value");
  }
}


_tools/meta.ts
import { BuildOptions } from "https://deno.land/x/dnt@0.33.1/mod.ts";

export const makeOptions = (version: string): BuildOptions => ({
  test: false,
  shims: {},
  compilerOptions: {
    lib: ["esnext", "dom"],
  },
  typeCheck: true,
  entryPoints: ["./mod.ts"],
  outDir: "./npm",
  package: {
    name: "@httpland/http-cache-control",
    version,
    description:
      "HTTP Cache-Control header middleware for standard request and response",
    keywords: [
      "http",
      "middleware",
      "cache-control",
      "handler",
      "request",
      "response",
    ],
    license: "MIT",
    homepage: "https://github.com/httpland/http-cache-control",
    repository: {
      type: "git",
      url: "git+https://github.com/httpland/http-cache-control.git",
    },
    bugs: {
      url: "https://github.com/httpland/http-cache-control/issues",
    },
    sideEffects: false,
    type: "module",
    publishConfig: {
      access: "public",
    },
  },
  packageManager: "pnpm",
});


cache_control.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

import { CachingHeader } from "./deps.ts";

export interface CacheDirectives {
  /**
   * @see [RFC 9111, 5.2.2.1. max-age](https://www.rfc-editor.org/rfc/rfc9111.html#name-max-age-2)
   */
  readonly "max-age"?: number;

  /**
   * @see [RFC 9111, 5.2.2.2. must-revalidate](https://www.rfc-editor.org/rfc/rfc9111.html#name-must-revalidate)
   */
  readonly "must-revalidate"?: true;

  /**
   * @see [RFC 9111, 5.2.2.3. must-understand](https://www.rfc-editor.org/rfc/rfc9111.html#name-must-understand)
   */
  readonly "must-understand"?: true;

  /**
   * @see [RFC 9111, 5.2.2.4. no-cache](https://www.rfc-editor.org/rfc/rfc9111.html#name-no-cache-2)
   */
  readonly "no-cache"?: true;

  /**
   * @see [5.2.2.5. no-store](https://www.rfc-editor.org/rfc/rfc9111.html#name-no-store-2)
   */
  readonly "no-store"?: true;

  /**
   * @see [RFC 9111, 5.2.2.6. no-transform](https://www.rfc-editor.org/rfc/rfc9111.html#name-no-transform-2)
   */
  readonly "no-transform"?: true;

  /**
   * @see [RFC 9111, 5.2.2.7. private](https://www.rfc-editor.org/rfc/rfc9111.html#name-private)
   */
  readonly private?: true;

  /**
   * @see [RFC 9111, 5.2.2.8. proxy-revalidate](https://www.rfc-editor.org/rfc/rfc9111.html#name-proxy-revalidate)
   */
  readonly "proxy-revalidate"?: true;

  /**
   * @see [RFC 9111, 5.2.2.9. public](https://www.rfc-editor.org/rfc/rfc9111.html#section-5.2.2.9)
   */
  readonly public?: true;

  /**
   * @see [RFC 9111, 5.2.2.10. s-maxage](https://www.rfc-editor.org/rfc/rfc9111.html#name-s-maxage)
   */
  readonly "s-maxage"?: number;
}

export interface ExtensionDirectives {
  /**
   * @see [RFC 8246, HTTP Immutable Responses](https://www.rfc-editor.org/rfc/rfc8246.html)
   */
  readonly immutable?: true;

  /** The `stale-while-revalidate` Cache-Control extension indicates that caches MAY serve the response in which it appears after it becomes stale,
   * up to the indicated number of seconds.
   * @see [RFC 5861, 3.The stale-while-revalidate Cache-Control Extension](https://www.rfc-editor.org/rfc/rfc5861#section-3)
   */
  readonly "stale-while-revalidate"?: number;

  /** The `stale-if-error` Cache-Control extension indicates that when an error is encountered,
   * a cached stale response MAY be used to satisfy the request,
   * regardless of other freshness information.
   * @see [RFC 5861, 4.The stale-if-error Cache-Control Extension](https://www.rfc-editor.org/rfc/rfc5861#section-4)
   */
  readonly "stale-if-error"?: number;
}

export function withCacheControl(
  _: Request,
  response: Response,
  directives: CacheDirectives & ExtensionDirectives,
): Response | Promise<Response> {
  const hasCacheControl = response.headers.has(CachingHeader.CacheControl);

  if (hasCacheControl) return response;

  const cacheControl = stringifyCacheDirectives({ ...directives });

  if (!cacheControl) return response;

  response.headers.set(CachingHeader.CacheControl, cacheControl);

  return response;
}

export function stringifyCacheDirectives(
  record: Record<string, number | true>,
): string {
  const fieldValues = Object.entries(record).map(([key, value]) =>
    stringifyDirective(key, value)
  );

  return fieldValues
    .filter(Boolean)
    .join(", ");
}

export function stringifyDirective(
  key: string,
  value: number | true,
): string {
  if (!key) return "";

  if (value === true) return key;

  return `${key}=${value}`;
}
